File: Comparative-Analysis/matrix_profiled.go
package main

import (
	"fmt"
	"math/rand"
	"os"
	"runtime"
	"runtime/pprof"
	"time"
)

func matrixMultiply(n int) {
	A := make([][]float64, n)
	B := make([][]float64, n)
	C := make([][]float64, n)
	for i := range A {
		A[i] = make([]float64, n)
		B[i] = make([]float64, n)
		C[i] = make([]float64, n)
		for j := range A[i] {
			A[i][j] = rand.Float64()
			B[i][j] = rand.Float64()
		}
	}
	for i := 0; i < n; i++ {
		for j := 0; j < n; j++ {
			for k := 0; k < n; k++ {
				C[i][j] += A[i][k] * B[k][j]
			}
		}
	}
}
func main() {
	f, _ := os.Create("cpu.prof")
	pprof.StartCPUProfile(f)
	defer pprof.StopCPUProfile()
	fmt.Printf("Number of CPUs: %d\n", runtime.NumCPU())
	fmt.Printf("Go Routines: %d\n", runtime.NumGoroutine())
	fmt.Printf("Go Version: %s\n", runtime.Version())
	start := time.Now()
	matrixMultiply(100)
	fmt.Println("Execution Time:", time.Since(start).Seconds(), "seconds")
}

File: Comparative-Analysis/matrix_profiled.py
import random
import time
import cProfile
from memory_profiler import profile
@profile
def matrix_multiply(n):
    A = [[random.random() for _ in range(n)] for _ in range(n)]
    B = [[random.random() for _ in range(n)] for _ in range(n)]
    result = [[0.0 for _ in range(n)] for _ in range(n)]
    for i in range(n):
        for j in range(n):
            for k in range(n):
                result[i][j] += A[i][k] * B[k][j]
    return result
if __name__ == "__main__":
    start = time.time()
    cProfile.run("matrix_multiply(100)")
    print(f"Execution Time: {time.time() - start:.2f} seconds")
